File: backend/app/services/group_discussion_service.py
import json
import logging
import re
import uuid
from datetime import datetime
from typing import List, Dict, Optional
from app.schemas.group_discussion import (
    GroupDiscussionStartRequest,
    GroupDiscussionStartResponse,
    GroupDiscussionResponseRequest,
    GroupDiscussionResponseResponse,
    GroupDiscussionEvaluation,
    GroupDiscussionSessionSummary,
    AIResponseData
)
from app.services.gemini_service import gemini_service
from app.services.prompts.group_discussion import (
    TOPIC_GENERATION_PROMPT,
    MODERATOR_INTRO_PROMPT,
    PARTICIPANT_RESPONSE_PROMPT,
    GD_EVALUATION_PROMPT
)

logger = logging.getLogger(__name__)

class GroupDiscussionService:
    async def start_session(self, request: GroupDiscussionStartRequest) -> GroupDiscussionStartResponse:
        session_id = f"gd_{uuid.uuid4().hex[:10]}"
        
        # 1. Generate topic
        topic_prompt = TOPIC_GENERATION_PROMPT.format(category=request.category, difficulty=request.difficulty)
        topic = await gemini_service.generate_communication_response(
            message=topic_prompt,
            mode="group_discussion"
        )
        if not topic or len(topic) < 5:
            topic = "Should artificial intelligence replace some human jobs?"
            
        topic = topic.replace('"', '').strip()

        # 2. Generate moderator intro
        mod_prompt = MODERATOR_INTRO_PROMPT.format(topic=topic)
        moderator_intro = await gemini_service.generate_communication_response(
            message=mod_prompt,
            mode="group_discussion"
        )
        
        if not moderator_intro:
            moderator_intro = f"Welcome everyone. Today's topic is: '{topic}'. Let's begin our discussion. Participant A, what are your initial thoughts?"

        return GroupDiscussionStartResponse(
            session_id=session_id,
            topic=topic,
            category=request.category,
            difficulty=request.difficulty,
            moderator_intro=moderator_intro
        )

    async def evaluate_turn(self, request: GroupDiscussionResponseRequest, topic: str) -> GroupDiscussionResponseResponse:
        # Generate AI participant responses
        conversation_history = "\n".join(
            [f"{msg.get('role', 'unknown')}: {msg.get('content', '')}" for msg in request.messages]
        )
        
        prompt = PARTICIPANT_RESPONSE_PROMPT.format(
            topic=topic,
            round_num=request.current_round,
            conversation_history=conversation_history,
            student_message=request.student_message
        )
        
        raw_response = await gemini_service.generate_communication_response(
            message=prompt,
            mode="group_discussion"
        )
        
        ai_responses = self._parse_participant_responses(raw_response)
        
        # Ensure fallback
        if not ai_responses:
            ai_responses = [AIResponseData(speaker="participant_A", content="That's an interesting point. I'd like to add that we should also consider the broader implications.")]
            
        is_complete = request.current_round >= 5
        
        return GroupDiscussionResponseResponse(
            session_id=request.session_id,
            round=request.current_round,
            ai_responses=ai_responses,
            is_complete=is_complete
        )
        
    async def complete_session(self, session_id: str, topic: str, category: str, difficulty: str, transcript_messages: List[Dict]) -> GroupDiscussionSessionSummary:
        # Full evaluation
        transcript = "\n".join(
            [f"{msg.get('role', 'unknown')}: {msg.get('content', '')}" for msg in transcript_messages]
        )
        
        eval_prompt = GD_EVALUATION_PROMPT.format(
            topic=topic,
            transcript=transcript
        )
        
        raw_eval = await gemini_service.generate_communication_response(
            message=eval_prompt,
            mode="group_discussion"
        )
        
        evaluation = self._parse_evaluation(raw_eval)
        
        return GroupDiscussionSessionSummary(
            session_id=session_id,
            category=category,
            difficulty=difficulty,
            topic=topic,
            evaluation=evaluation,
            completed_at=datetime.utcnow().isoformat() + "Z"
        )
        
    def _parse_participant_responses(self, raw_response: str) -> List[AIResponseData]:
        try:
            clean_json_str = raw_response
            if "```json" in raw_response:
                clean_json_str = raw_response.split("```json")[1].split("```")[0].strip()
            elif "```" in raw_response:
                clean_json_str = raw_response.split("```")[1].split("```")[0].strip()
                
            data = json.loads(clean_json_str)
            responses = []
            for item in data:
                responses.append(AIResponseData(speaker=item["speaker"], content=item["content"]))
            return responses
        except Exception as e:
            logger.warning(f"Failed to parse participant responses JSON: {e}")
            return []
            
    def _parse_evaluation(self, raw_eval: str) -> GroupDiscussionEvaluation:
        try:
            clean_json_str = raw_eval
            if "```json" in raw_eval:
                clean_json_str = raw_eval.split("```json")[1].split("```")[0].strip()
            elif "```" in raw_eval:
                clean_json_str = raw_eval.split("```")[1].split("```")[0].strip()
                
            data = json.loads(clean_json_str)
            return GroupDiscussionEvaluation(**data)
        except Exception as e:
            logger.warning(f"Failed to parse evaluation JSON: {e}. Returning fallback.")
            return GroupDiscussionEvaluation(
                communication=75, clarity=75, grammar=75, vocabulary=75, relevance=75,
                confidence=75, participation=75, leadership=75, teamwork=75,
                respectfulness=75, argumentQuality=75, responsiveness=75,
                adaptability=75, timeManagement=75, overallScore=75,
                strengths=["Participated in discussion"],
                improvements=["Could be more active"],
                improved_responses=[]
            )

group_discussion_service = GroupDiscussionService()

File: frontend/src/services/groupDiscussionService.ts
import {
  doc,
  setDoc,
  collection,
  query,
  where,
  orderBy,
  getDocs,
  serverTimestamp,
  updateDoc
} from 'firebase/firestore';
import { db } from './firebase';
import { ApiService } from './api';
import { recordStudentActivity } from './progressService';

export interface AIResponseData {
  speaker: string;
  content: string;
}

export interface GroupDiscussionEvaluation {
  communication: number;
  clarity: number;
  grammar: number;
  vocabulary: number;
  relevance: number;
  confidence: number;
  participation: number;
  leadership: number;
  teamwork: number;
  respectfulness: number;
  argumentQuality: number;
  responsiveness: number;
  adaptability: number;
  timeManagement: number;
  overallScore: number;
  strengths: string[];
  improvements: string[];
  improved_responses?: { original: string; improved: string; reason: string }[];
}

export interface GroupDiscussionStartResponse {
  session_id: string;
  topic: string;
  category: string;
  difficulty: string;
  moderator_intro: string;
}

export interface GroupDiscussionRespondResponse {
  session_id: string;
  round: number;
  ai_responses: AIResponseData[];
  is_complete: boolean;
}

export interface GroupDiscussionSummary {
  session_id: string;
  category: string;
  difficulty: string;
  topic: string;
  evaluation: GroupDiscussionEvaluation;
  completed_at: string;
}

export interface GroupDiscussionHistoryItem {
  sessionId: string;
  dateStr: string;
  difficulty: string;
  category: string;
  topic: string;
  score: number;
  status: 'active' | 'completed';
}

export const startGroupDiscussion = async (
  uid: string,
  category: string,
  difficulty: string
): Promise<GroupDiscussionStartResponse> => {
  const data = await ApiService.post<GroupDiscussionStartResponse>(
    '/communication/group-discussion/start',
    { uid, category, difficulty }
  );

  const sessionRef = doc(db!, `users/${uid}/communicationSessions/${data.session_id}`);
  await setDoc(sessionRef, {
    mode: 'group_discussion',
    category,
    difficulty,
    topic: data.topic,
    status: 'active',
    round: 1,
    startedAt: serverTimestamp(),
    lastActivityAt: serverTimestamp(),
  });

  return data;
};

export const respondGroupDiscussion = async (
  uid: string,
  sessionId: string,
  studentMessage: string,
  currentRound: number,
  topic: string,
  messages: any[]
): Promise<GroupDiscussionRespondResponse> => {
  const data = await ApiService.post<GroupDiscussionRespondResponse>(
    '/communication/group-discussion/respond',
    {
      uid,
      session_id: sessionId,
      student_message: studentMessage,
      current_round: currentRound,
      topic,
      messages,
    }
  );

  const sessionRef = doc(db!, `users/${uid}/communicationSessions/${sessionId}`);
  await updateDoc(sessionRef, {
    round: currentRound + 1,
    lastActivityAt: serverTimestamp(),
  });

  return data;
};

export const completeGroupDiscussion = async (
  uid: string,
  sessionId: string,
  topic: string,
  category: string,
  difficulty: string,
  messages: any[]
): Promise<GroupDiscussionSummary> => {
  const data = await ApiService.post<GroupDiscussionSummary>(
    '/communication/group-discussion/complete',
    {
      session_id: sessionId,
      topic,
      category,
      difficulty,
      messages,
    }
  );

  const sessionRef = doc(db!, `users/${uid}/communicationSessions/${sessionId}`);
  await updateDoc(sessionRef, {
    status: 'completed',
    completedAt: serverTimestamp(),
    evaluation: data.evaluation,
    overallScore: data.evaluation.overallScore,
  });

  await recordStudentActivity(uid, {
    module: 'communication',
    activityType: 'group_discussion_session' as any,
    status: 'completed',
    score: data.evaluation.overallScore,
  });

  return data;
};

export const getGroupDiscussionHistory = async (uid: string): Promise<GroupDiscussionHistoryItem[]> => {
  const sessionsRef = collection(db!, `users/${uid}/communicationSessions`);
  const q = query(
    sessionsRef,
    where('mode', '==', 'group_discussion'),
    orderBy('startedAt', 'desc')
  );

  const snapshot = await getDocs(q);
  const history: GroupDiscussionHistoryItem[] = [];

  snapshot.forEach((docSnap) => {
    const data = docSnap.data();
    const dateStr = data.startedAt?.toDate().toLocaleDateString() || new Date().toLocaleDateString();

    history.push({
      sessionId: docSnap.id,
      dateStr,
      difficulty: data.difficulty || 'medium',
      category: data.category || 'technology',
      topic: data.topic || 'Unknown Topic',
      score: data.overallScore || 0,
      status: data.status as 'active' | 'completed',
    });
  });

  return history;
};
